### PR TITLE
Refs [Task][30718] Apply MVVM Structure

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 30718: Apply MVVM Structure](https://edu-redmine.sun-asterisk.vn/issues/30718)

## WHAT
- Apply MVVM Structure
## HOW
- Apply MVVM Structure
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -34,4 +42,20 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.2'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
 
+    // Material Design
+    implementation 'com.google.android.material:material:1.2.1'
+
+    // Architecture Component
+    implementation "androidx.lifecycle:lifecycle-extensions:$versions.lifecycleVersion"
+    implementation "androidx.lifecycle:lifecycle-livedata-ktx:$versions.lifecycleVersion"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:$versions.lifecycleVersion"
```
Comment (thonc-2135) [inline app/build.gradle:51]: mình tưởng là dùng lifecycle extension là n có sẵn viewmodelv s livedata r nhỉ, bạn thử check xem phải k

File: app/src/main/java/com/sunasterisk/appchat/db/Result.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.sunasterisk.appchat.db
+
+sealed class Result<out T : Any> {
+
+    data class Success<out T : Any>(val data: T) : Result<T>()
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/appchat/db/Result.kt:5]: dùng out vs k dùng out khác gì nhau k ?

Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/appchat/db/Result.kt:5]: dùng out (generic covariant type) mở rộng hơn 1 số khả năng so với ko dùng out (invariant generic):
+ polymorphism: có thể dùng sub type thay cho super type
+ giới hạn chỉ sử dụng nó cho return type hoặc val properties

File: app/build.gradle (hunk 1 above)
Comment (hieuna-441907) [reply app/build.gradle:51]: Ok ông, đúng là có sẵn, nhưng nó bị deprecated từ API level 1.1.0, nên dùng 2 cái kia để thay thế

**Review by NguyenMinhTuanB — APPROVED** [review]

**Pull request merged**